### PR TITLE
XDR XQL Query Engine Integration

### PR DESCRIPTION
**Pull request by dansterenson**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/39474

## Description
XDR XQL query engine integration
Cortex_XDR_-_XQL_Thu_Sep_02_2021

## Minimum version of Cortex XSOAR
- [x] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by dansterenson** [comment]
Ready for review, validation will pass when the new SDK release is out. The only thing that has not yet been decided is which arguments should be marked 'required' in the built in queries (waiting for a response from Elior).

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+        'request_data': {
+            'stream_id': stream_id,
+            'is_gzip_compressed': True,
+        }
+    }
+    # Call the Client function and get the raw response
+    return client.get_query_result_stream(data)
+
+
+def format_results(list_to_format: list, remove_empty_fields: bool = True) -> list:
```
Comment (teizenman) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:471]: This is a pretty messy function.
Think of a different implementation for it, and look at CommonServerPython to see if we have any relevant functions there.

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 2]
```diff
@@ -0,0 +1,767 @@
...
+            'day': 24 * 60 * 60 * 1000,
+            'week': 7 * 24 * 60 * 60 * 1000,
+            'month': 30 * 24 * 60 * 60 * 1000,
+            'year': 12 * 30 * 24 * 60 * 60 * 1000
+        }
+        for k in time_multiples.keys():
+            if k in time_to_convert:
+                num = int(time_to_convert.split(' ')[0])
+                return time_multiples[k] * num
+        raise ValueError
```
Comment (teizenman) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:367]: Why not just raise the DemistoException here?

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 3]
```diff
@@ -0,0 +1,767 @@
...
+        raise DemistoException('Please provide a process_causality_id argument.')
+    process_causality_id_list = format_arg(process_causality_id_list)
+
+    return f'dataset = xdr_data | filter agent_id in ({endpoint_ids}) and event_type = NETWORK and ' \
+           f'actor_process_causality_id in ({process_causality_id_list}) | fields agent_hostname, agent_ip_addresses, ' \
+           f'agent_id, action_local_ip, action_remote_ip, action_remote_port, dst_action_external_hostname, ' \
+           f'dns_query_name, action_app_id_transitions, action_total_download, action_total_upload, action_country, ' \
+           f'action_as_data, actor_process_image_sha256, actor_process_image_name , actor_process_image_path, ' \
+           f'actor_process_signature_vendor, actor_process_signature_product, actor_causality_id, ' \
+           f'actor_process_image_command_line, actor_process_instance_id'
```
Comment (teizenman) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:325]:
```suggestion
    return f'''dataset = xdr_data | filter agent_id in ({endpoint_ids}) and event_type = NETWORK and 
           actor_process_causality_id in ({process_causality_id_list}) | fields agent_hostname, agent_ip_addresses, 
           agent_id, action_local_ip, action_remote_ip, action_remote_port, dst_action_external_hostname, 
           dns_query_name, action_app_id_transitions, action_total_download, action_total_upload, action_country, 
           action_as_data, actor_process_image_sha256, actor_process_image_name , actor_process_image_path, 
           actor_process_signature_vendor, actor_process_signature_product, actor_causality_id, 
           actor_process_image_command_line, actor_process_instance_id'''
```

**Review by teizenman — CHANGES_REQUESTED** [review]
I only went through the basic stuff and the code itself.
1. Please change the docstrings to google style.
2. Some of the comments apply to more than one place, fix all of the relevant comments in any relevant place.

File: Packs/CortexXDR/pack_metadata.json [hunk 4]
```diff
@@ -2,7 +2,7 @@
     "name": "Palo Alto Networks Cortex XDR - Investigation and Response",
     "description": "Automates Cortex XDR incident response, and includes custom Cortex XDR incident views and layouts to aid analyst investigations.",
     "support": "xsoar",
-    "currentVersion": "3.0.24",
+    "currentVersion": "3.0.25",
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/pack_metadata.json:5]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 5]
```diff
@@ -0,0 +1,767 @@
...
+    params = demisto.params()
+    api_key = params.get('apikey')
+    api_key_id = params.get('apikey_id')
+    base_url = urljoin(params['url'], '/public_api/v1')
+    verify_cert = not params.get('insecure', False)
+    proxy = params.get('proxy', False)
+    command = demisto.command()
+    demisto.debug(f'Command being called is {command}')
+    try:
+        nonce = "".join([secrets.choice(string.ascii_letters + string.digits) for _ in range(64)])
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:730]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 6]
```diff
@@ -0,0 +1,767 @@
...
+        timestamp = str(int(datetime.now(timezone.utc).timestamp()) * 1000)
+        auth_key = "%s%s%s" % (api_key, nonce, timestamp)
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:732]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+        api_key_hash = hashlib.sha256(auth_key.encode("utf-8")).hexdigest()
+
+        headers = {
+            "x-xdr-timestamp": str(timestamp),
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:736]: Right :+1:

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+            "x-xdr-nonce": nonce,
+            "x-xdr-auth-id": str(api_key_id),
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:738]: :+1:

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 9]
```diff
@@ -0,0 +1,767 @@
...
+            "Authorization": api_key_hash
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:739]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 10]
```diff
@@ -0,0 +1,767 @@
...
+        }
+
+        client = Client(
+            base_url=base_url,
+            verify=verify_cert,
+            headers=headers,
+            proxy=proxy)
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:746]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 11]
```diff
@@ -0,0 +1,767 @@
...
+            return_results(get_built_in_query_results_polling_command(client, args))
+        else:
+            raise NotImplementedError(f'Command {command} does not exist.')
+
+    # Log exceptions and return errors
+    except Exception as e:
+        demisto.error(traceback.format_exc())  # print the traceback
+        return_error(f'Failed to execute {command} command.\nError: {str(e)}')
+    finally:
+        get_integration_context().clear()
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:761]: You are right, basically I want to clear the integration context in two of the following scenarios:
1. Polling status is "SUCCESS" and polling is over.
2. An error occurred while the schedule commands are still running.
I made a mistake and using finally here is not right, I used set_integration_context({}) instead and added it to the exception scope.

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 12]
```diff
@@ -0,0 +1,767 @@
...
+
+    :type str_to_format: ``str``
+    :param str_to_format: The string input to format.
+
+    :return: The formatted string.
+    :rtype: ``str``
+    """
+    if not str_to_format:
+        str_to_format = ''
+    return ','.join(map('"{0}"'.format, argToList(str_to_format)))
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:91]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 13]
```diff
@@ -0,0 +1,767 @@
...
+    """
+    # build query, if no endpoint_id was given, the query will search in every endpoint_id (*).
+    endpoint_id_list = format_arg(args.get('endpoint_id', '*'))
+    available_commands = init_built_in_commands()
+    query = available_commands.get(demisto.command(), {}).get('func')(endpoint_id_list, args)
+
+    # add extra fields to query
+    extra_fields_list = ", ".join(str(e) for e in argToList(args.get('extra_fields', [])))
+    extra_fields_list = f', {extra_fields_list}' if extra_fields_list else ''  # add comma to the beginning of fields
+    query = f'{query}{extra_fields_list}'
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:697]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/README.md [hunk 14]
```diff
@@ -0,0 +1,5 @@
+This README contains the full documentation for your integration.
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/README.md]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 15]
```diff
@@ -310,22 +302,11 @@ def convert_relative_time_to_milliseconds(time_to_convert: str) -> int:
...
-                num = int(time_to_convert.split(' ')[0])
-                return time_multiples[k] * num
-        raise ValueError
-    except ValueError:
-        raise DemistoException('Please enter a valid time frame (seconds, minutes, hours, days, weeks, months, years).')
+        relative = dateparser.parse(time_to_convert, settings={'TIMEZONE': 'UTC'})
+        return int((datetime.utcnow() - relative).total_seconds()) * 1000
+    except Exception as exc:
+        raise DemistoException(f'Please enter a valid time frame (seconds, minutes, hours, days, weeks, months, '
+                               f'years).\nstr({exc})')
```
Comment (teizenman) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:309]:
```suggestion
                               f'years).\n{str(exc)}')
```

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 16]
```diff
@@ -620,27 +620,30 @@ def get_xql_query_results_polling_command(client: Client, args: dict) -> Union[C
     command_data = integration_context.get(query_id, args)
     command_name = command_data.get('command_name', demisto.command())
     interval_in_secs = int(args.get('interval_in_seconds', 10))
+    max_fields = int(args.get('max_fields', 20))
```
Comment (teizenman) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:623]: It's better to use `arg_to_int` here.

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 17]
```diff
@@ -785,8 +793,10 @@ def main() -> None:
     """
     args = demisto.args()
     params = demisto.params()
-    api_key = params.get('apikey')
-    api_key_id = params.get('apikey_id')
+    if not params.get('apikey') or not (api_key := params.get('apikey', {}).get('password')):
```
Comment (teizenman) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:796]: Why are you using the `:=` operator here?

File: Tests/conf.json [hunk 18]
```diff
@@ -4517,6 +4517,11 @@
             "integrations": [
                 "TOPdesk"
             ]
+        },
+        {
+            "integrations": "Cortex XDR - XQL Query Engine",
+            "playbookID": "Cortex XDR - XQL Query - Test",
+            "fromversion": "5.0.0"
```
Comment (teizenman) [inline Tests/conf.json:4524]:
```suggestion
            "fromversion": "6.2.0"
```

**Review by teizenman — CHANGES_REQUESTED** [review]

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py (hunk 16 above)
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:623]: Done

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py (hunk 17 above)
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:796]: Fixed

File: Tests/conf.json (hunk 18 above)
Comment (dansterenson) [reply Tests/conf.json:4524]: keeping 5.0.0 as we spoked.

**Review by teizenman — APPROVED** [review]
Fix final things we discussed

**Comment by dansterenson** [comment]
@ShirleyDenkberg Can you please go over the README file? :)

**Comment by ShirleyDenkberg** [comment]
@dansterenson I edited the integration yml file.  Please regenerate the integration README.
@teizenman @yaakovi Doc review completed.

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/test_data/baseintegration-dummy.json [hunk 19]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/test_data/baseintegration-dummy.json:1]: should be removed

File: Packs/CortexXDR/pack_metadata.json [hunk 20]
```diff
@@ -2,7 +2,7 @@
     "name": "Palo Alto Networks Cortex XDR - Investigation and Response",
     "description": "Automates Cortex XDR incident response, and includes custom Cortex XDR incident views and layouts to aid analyst investigations.",
     "support": "xsoar",
-    "currentVersion": "3.0.27",
+    "currentVersion": "3.1.0",
```
Comment (yaakovi) [inline Packs/CortexXDR/pack_metadata.json:5]: new integration should bump version by major

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py [hunk 21]
```diff
@@ -0,0 +1,849 @@
...
+    'xdr-xql-get-quota': get_xql_quota_command,
+}
+
+
+def main() -> None:
+    """main function, parses params and runs command functions
+    """
+    args = demisto.args()
+    params = demisto.params()
+    apikey = params.get('apikey', {}).get('password', '')
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py:798]:
```suggestion
    # using two different credentials object as they both fields need to be encrypted
    apikey = params.get('apikey', {}).get('password', '')
```

**Review by yaakovi — APPROVED** [review]

File: Packs/CortexXDR/Integrations/XQLQueryingEngine/test_data/baseintegration-dummy.json (hunk 19 above)
Comment (dansterenson) [reply Packs/CortexXDR/Integrations/XQLQueryingEngine/test_data/baseintegration-dummy.json:1]: Done

File: Packs/CortexXDR/pack_metadata.json (hunk 20 above)
Comment (dansterenson) [reply Packs/CortexXDR/pack_metadata.json]: Done

**Pull request merged**